cli/cmd/addon_publish.go

```diff
@@ -0,0 +1,245 @@
+package cmd
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"syscall"
+
+	cloudquery_api "github.com/cloudquery/cloudquery-api-go"
+	"github.com/cloudquery/cloudquery/cli/internal/auth"
+	"github.com/spf13/cobra"
+)
+
+const (
+	addonPublishShort = "Publish to CloudQuery Hub."
+	addonPublishLong  = `Publish to CloudQuery Hub.
+
+This publishes an addon version to CloudQuery Hub from a manifest file.
+`
+	addonPublishExample = `
+# Publish an addon version from a manifest file
+cloudquery addon publish /path/to/manifest.json v1.0.0`
+)
+
+func newCmdAddonPublish() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "publish manifest.json v1.0.0 [--finalize]",
+		Short:   addonPublishShort,
+		Long:    addonPublishLong,
+		Example: addonPublishExample,
+		Hidden:  true,
+		Args:    cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// Set up a channel to listen for OS signals for graceful shutdown.
+			ctx, cancel := context.WithCancel(cmd.Context())
+
+			sigChan := make(chan os.Signal, 1)
+			signal.Notify(sigChan, syscall.SIGTERM)
+
+			go func() {
+				<-sigChan
+				cancel()
+			}()
+
+			return runAddonPublish(ctx, cmd, args)
+		},
+	}
+	cmd.Flags().BoolP("finalize", "f", false, `Finalize the addon version after publishing. If false, the addon version will be marked as draft=true.`)
+
+	return cmd
+}
+
+type ManifestJSONV1 struct {
+	Type      string `json:"type"` // always "addon"
+	TeamName  string `json:"team_name"`
+	AddonName string `json:"addon_name"`
+	// AddonType   string   `json:"addon_type"`   // unused
+	// AddonFormat string   `json:"addon_format"` // unused
+
+	PathToMessage string `json:"message"`
+	PathToZip     string `json:"path"`
+	PathToDoc     string `json:"doc"`
+
+	PluginDeps []string `json:"plugin_deps"`
+	AddonDeps  []string `json:"addon_deps"`
+}
+
+func runAddonPublish(ctx context.Context, cmd *cobra.Command, args []string) error {
+	token, err := auth.GetToken()
+	if err != nil {
+		return fmt.Errorf("failed to get auth token: %w", err)
+	}
+
+	manifestPath, version := args[0], args[1]
+	manifestDir := filepath.Dir(manifestPath)
+
+	manifest, err := readManifestJSON(manifestPath)
+	if err != nil {
+		return fmt.Errorf("failed to read manifest.json: %w", err)
+	}
+
+	if manifest.Type != "addon" {
+		return errors.New("manifest.json is not of addon type")
+	}
+
+	zipPath := filepath.Join(manifestDir, manifest.PathToZip)
+	if _, err := os.Stat(zipPath); err != nil {
+		return fmt.Errorf("could not read file %s: %w", zipPath, err)
+	}
+
+	name := fmt.Sprintf("%s/%s@%s", manifest.TeamName, manifest.AddonName, version)
+	fmt.Printf("Publishing %s to CloudQuery Hub...\n", name)
+
+	c, err := cloudquery_api.NewClientWithResponses(getEnvOrDefault("CLOUDQUERY_API_URL", defaultAPIURL),
```

Comment:
Let's create a constant in `constants.go` for `CLOUDQUERY_API_URL`, since it's used in multiple places now

Reply:
Done in [`fcda030` (#14564)](https://github.com/cloudquery/cloudquery/pull/14564/commits/fcda03061b03e7560b91ffcf3eb53f60dcaa6eff)